### PR TITLE
GlobalException + validation

### PR DESCRIPTION
**Pull request by Eidenn005**
Added CustomGlobalExceptionHandler
Added validation for DTO

### CODE REVIEW

File: src/main/java/com/example/onlinebookstore/dto/CreateBookRequestDto.java [hunk 1]
```diff
@@ -1,13 +1,22 @@
...
+    @Min(0)
     private BigDecimal price;
+    @NotNull
+    @Min(5)
     private String author;
+    @NotNull
+    @Min(13)
     private String isbn;
     private String description;
     private String coverImage;
```
Comment (Bohadan-Markatov) [inline src/main/java/com/example/onlinebookstore/dto/CreateBookRequestDto.java]: I’m not sure that the min and max annotation is suitable for strings, there is a special annotation for this, also add a maximum length limit as in the table creation file for liquibase. you also need to create an annotation and a validator class for the isbn field

File: src/main/java/com/example/onlinebookstore/controller/BookController.java [hunk 2]
```diff
@@ -34,7 +35,7 @@ public BookDto getBookById(@PathVariable Long id) {
 
     @ResponseStatus(HttpStatus.CREATED)
     @PostMapping
-    public BookDto createBook(@RequestBody CreateBookRequestDto bookDto) {
+    public BookDto createBook(@RequestBody @Valid CreateBookRequestDto bookDto) {
         return bookService.save(bookDto);
     }
```
Comment (Bohadan-Markatov) [inline src/main/java/com/example/onlinebookstore/controller/BookController.java:41]: if you have update method add also valid annotation there

**Review by Bohadan-Markatov — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/onlinebookstore/dto/CreateBookRequestDto.java (hunk 1 above)
Comment (Eidenn005) [reply src/main/java/com/example/onlinebookstore/dto/CreateBookRequestDto.java]: Add @Size annos.
Change some with NotBlank.

File: src/main/java/com/example/onlinebookstore/controller/BookController.java (hunk 2 above)
Comment (Eidenn005) [reply src/main/java/com/example/onlinebookstore/controller/BookController.java:41]: done

File: src/main/java/com/example/onlinebookstore/dto/CreateBookRequestDto.java [hunk 3]
```diff
@@ -1,13 +1,24 @@
...
     private String title;
+    @NotNull
+    @Min(0)
     private BigDecimal price;
+    @NotBlank
+    @Size(min = 5, max = 100)
     private String author;
+    @NotNull
+    @Size(min = 13)
     private String isbn;
```
Comment (andrii-hoienko) [inline src/main/java/com/example/onlinebookstore/dto/CreateBookRequestDto.java]:
```suggestion
    @ISBN
    private String isbn;
```

File: src/main/java/com/example/onlinebookstore/exception/CustomGlobalExceptionHandler.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    private String getErrorMessage(ObjectError e) {
+        if (e instanceof FieldError) {
+            String field = ((FieldError) e).getField();
+            String message = e.getDefaultMessage();
+            return field + " " + message;
+        }
+        return e.getDefaultMessage();
+    }
```
Comment (andrii-hoienko) [inline src/main/java/com/example/onlinebookstore/exception/CustomGlobalExceptionHandler.java]: Please add handling of other error messages that can occur in code (EntityNotFoundException) and generic error as well (Exception)

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]
Nice, left minor comments

File: src/main/java/com/example/onlinebookstore/dto/CreateBookRequestDto.java (hunk 3 above)
Comment (Eidenn005) [reply src/main/java/com/example/onlinebookstore/dto/CreateBookRequestDto.java]: done

File: src/main/java/com/example/onlinebookstore/exception/CustomGlobalExceptionHandler.java (hunk 4 above)
Comment (Eidenn005) [reply src/main/java/com/example/onlinebookstore/exception/CustomGlobalExceptionHandler.java]: done

File: src/main/java/com/example/onlinebookstore/exception/CustomGlobalExceptionHandler.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+            Exception ex,
+            WebRequest request
+    ) {
+        Map<String, Object> body = new LinkedHashMap<>();
+        body.put("timestamp", LocalDateTime.now());
+        body.put("status", HttpStatus.INTERNAL_SERVER_ERROR);
+        body.put("error", "An unexpected error occurred: " + ex.getMessage());
+
+        return new ResponseEntity<>(body, HttpStatus.INTERNAL_SERVER_ERROR);
+    }
```
Comment (andrii-hoienko) [inline src/main/java/com/example/onlinebookstore/exception/CustomGlobalExceptionHandler.java]: Code duplication, I suggest to create a method that creates `Map<String, Object> body = new LinkedHashMap<>();` and accepts message and status code, also `timestamp, status, error` could be constant

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]
Nice, left one comment

File: src/main/java/com/example/onlinebookstore/exception/CustomGlobalExceptionHandler.java (hunk 5 above)
Comment (Eidenn005) [reply src/main/java/com/example/onlinebookstore/exception/CustomGlobalExceptionHandler.java]: Done

File: src/main/java/com/example/onlinebookstore/exception/CustomGlobalExceptionHandler.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+        return e.getDefaultMessage();
+    }
+
+    private ResponseEntity<Object> buildResponseEntity(HttpStatus status, String message) {
+        Map<String, Object> body = new LinkedHashMap<>();
+        body.put(TIMESTAMP, LocalDateTime.now());
+        body.put(STATUS, status);
+        body.put(ERRORS, message);
+        return new ResponseEntity<>(body, status);
+    }
```
Comment (andrii-hoienko) [inline src/main/java/com/example/onlinebookstore/exception/CustomGlobalExceptionHandler.java]:
```suggestion
    @Override
    protected ResponseEntity<Object> handleMethodArgumentNotValid(
            MethodArgumentNotValidException ex,
            HttpHeaders headers,
            HttpStatusCode status,
            WebRequest request
    ) {
        List<String> errors = ex.getBindingResult().getAllErrors().stream()
                .map(this::getErrorMessage)
                .toList();
        return buildResponseEntity(BAD_REQUEST, errors)
    }

    @ExceptionHandler(EntityNotFoundException.class)
    protected ResponseEntity<Object> handleEntityNotFoundException(
            EntityNotFoundException ex,
            WebRequest request
    ) {
        return buildResponseEntity(HttpStatus.NOT_FOUND, ex.getMessage());
    }

    @ExceptionHandler(Exception.class)
    protected ResponseEntity<Object> handleGenericException(
            Exception ex,
            WebRequest request
    ) {
        return buildResponseEntity(HttpStatus.INTERNAL_SERVER_ERROR,
                "An unexpected error occurred: " + ex.getMessage());
    }

    private String getErrorMessage(ObjectError e) {
        if (e instanceof FieldError) {
            String field = ((FieldError) e).getField();
            String message = e.getDefaultMessage();
            return field + " " + message;
        }
        return e.getDefaultMessage();
    }

    private ResponseEntity<Object> buildResponseEntity(HttpStatus status, Object message) { //or List<String>
        Map<String, Object> body = new LinkedHashMap<>();
        body.put(TIMESTAMP, LocalDateTime.now());
        body.put(STATUS, status);
        body.put(ERRORS, message);
        return new ResponseEntity<>(body, status);
    }
```

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]
Last charge

File: src/main/java/com/example/onlinebookstore/exception/CustomGlobalExceptionHandler.java (hunk 6 above)
Comment (Eidenn005) [reply src/main/java/com/example/onlinebookstore/exception/CustomGlobalExceptionHandler.java]: Done

**Review by andrii-hoienko — APPROVED** [review]
Good job!

**Pull request merged**